compiler/src/dotty/tools/dotc/core/Denotations.scala

```diff
@@ -585,8 +585,13 @@ object Denotations {
     final def signature(using Context): Signature =
       if (isType) Signature.NotAMethod // don't force info if this is a type SymDenotation
       else info match {
-        case info: MethodicType =>
-          try info.signature
+        case info: MethodOrPoly =>
+          val isJava =
+            if prefix eq NoPrefix then
+              symbol.is(JavaDefined)
+            else
+              prefix.classSymbol.is(JavaDefined)
```

Comment:
What if `prefix` is an `AndType` or `OrType` with some classes that are Java defined?

Reply:
Good question, I don't think this was ever handled when merging a JavaMethodType and a MethodType.

- For override checks, the prefix is always a ThisType where classSymbol delegates to its underlying which could be an AndType if a self-type is present, in that case we should fallback to ThisType#cls instead
- In other situation, what matter I think is that we respect this law:
   ```
   denot.prefix != NoPrefix
              implies
   denot   is equivalent to    denot.prefix.member(denot.name).atSignature(denot.signature)
   ```
   That is, whatever signature we choose must not be the signature of another overload, so for an intersection of all Java classes we need to pick the Java signature, otherwise we should be able to pick the Scala signature.

I'll add some test cases and tweak the implementation as needed.

Reply:
Actually we know that law is violated already, hence the existence of SELECTin, so as long as we consistently pick one or the other that might be good enough.